slapo/primitives/tensor_parallel.py

```diff
@@ -96,75 +70,16 @@ def _shard(name, tensor):
                 f"sharded along axis {sch.metadata.primitives['shard'][tensor_name]}"
             ) from None
 
-        def set_output_type(output_type, gather_axis=None):
-            try:
-                sch.metadata.primitives["shard"]["output_type"] = output_type
-            except KeyError:
-                raise RuntimeError(
-                    f"Output type of {sch.path} is already "
-                    f"{sch.metadata.primitives['shard']['output_type']}, but "
-                    f"{output_type} is requested"
-                ) from None
-
-            if gather_axis is not None:
-                try:
-                    sch.metadata.primitives["shard"]["gather_axis"] = gather_axis
-                except KeyError:
-                    raise RuntimeError(
-                        f"Output of {sch.path} has to be gathered along axis "
-                        f"{sch.metadata.primitives['shard']['gather_axis']}, but "
-                        f"{gather_axis} is requested"
-                    ) from None
-
-        out_type, out_part_axis = postproc_sharding(
-            sch.mod, tensor_name, sharded_size, axis
+        apply_shard_method("postproc", sch, tensor_name, sharded_size, axis)
```

Comment:
I feel like the new design makes what was easy a bit over complicated. 

1. What is the relationship between apply_shard_method and ShardMethod?
2. What is the api in ShardMethod such as postproc and sync used for? It seems to me these are in the old schedule's shard or sync API? And what is the relationship between this sync and old sync in schedule?

Reply:
I agree that the new design is not that straightforward, but I believe this reduces the possible ad-hoc changes in the future. For example, the previous PR #23 has to add many ad-hoc changes specifically for nn.Linear into `def shard`.

1. `apply_shard_method` determines which ShardMethod to be used. For example now we have LinearShardMethod, Conv2dShardMethod, etc. Thus, when `apply_shard_method("postproc", sch, ...)` and `sch.mod` is `nn.Linear`, we will dispatch to the sharding method of `nn.Linear`. This design lets the future developments of sharding method not change the schedule primitive directly.
2. The `postproc` is used to update the attributes (e.g., `linear.in_feature`) after sharding. This is renamed from `get_output_type_after_sharding`. The `sync` is migrated from the old sync primitive. By encapsulating this logic into ShardMethod, we are able to override it for a particular module. For example, when syncing `nn.Linear` with its input feature sharded, we have to apply all-reduce before bias addition, so we cannot simply register a hook. Instead, we replace `nn.Linear` with `LinearWithSyncFunc` and directly insert all_reduce in its forward, so that we could perform all-reduce before bias addition.